### DISCUSSION TITLE
More flexible value types?

### DISCUSSION THREAD

**Discussion by mansellan** [twinbasic/twinbasic#595] [Ideas]
VB.Net and C# are much more flexible when it comes to defining new value types. Where VBx has User Defined Types, VB.Net has `Structure` and C# has `struct`. These are value types, so:

1. Assignment to another variable creates a (shallow) copy
2. They _usually_ live on the stack
3. They are _usually_ passed by value
4. They have a number of restrictions as compared to classes:
  * Always have a parameterless constructor (usually synthesized)
  * No inheritance
  * Can implement interfaces (but that's often a bad idea because of boxing)
5. Can (and frequently should) implement operator overloads

Apart from those points (and a few more esoteric ones), they can use many of the same constructs - Subs, Functions, Properties, Fields, even Events and Delegates etc.

In .Net, they are usually worth considering for small ( < 16 byte) data models (which can also encapsulate some behavior), as they tend to be more lightweight due to usually being stack-allocated.

A good example would be Coordinates. A Structure could define x and y components, and perhaps a method to compute the angle from an origin.

In this context, VBx User Defined Types are like highly restricted versions of Structures - basically only allowing Fields (and at the same visibility as the UDT itself).

So this question is this? Is it worth adding something similar to twinBASIC? I honestly don't know the answer here.

The big win in .Net comes from reducing work for the Garbage Collector, which tB (rightly) won't ever have. But are there still advantages from being (mostly) stack allocated, even in a reference-counted setting?

**Comment by mansellan** [comment]
I guess a follow on question would be, does boxing/unboxing ever happen in VBx? @WaynePhillipsEA ? If not, it would definitely be worth never introducing it (by this suggestion or any other!)

**Comment by mwolfe02** [comment]
Part of my frustration with the limitations of UDTs in VBA is that class modules are "heavy" from a developer UI perspective. 

You can't use UDTs directly in a For Each loop, can't add them to a collection/dictionary, etc.  My typical workaround is to convert the UDT to a class module by just copying the UDT members into an empty class and making them public class variables.  This is not a big deal, except that now you have an extra class module taking up space in the single big list of Class Modules in the Project Explorer.  It also means that instead of keeping all related code for a standardized code module in one place, you have multiple files to export and import among VBA projects.

twinBASIC solves these problems by allowing you to define the class module in the same .twin file.  So now you can keep logically-related standard and class modules together in a single file.  Plus, you can organize code modules under a folder structure, which is another big win over VBA.

I guess this is my way of voting against this proposal.  It would add complexity to the compiler, but it would also add complexity for developers trying to decide what type they need in a given situation: should this be a UDT, or a struct, or a class module, (or a tuple...if those get implemented someday)?

That said, I don't really feel too strongly about it in either direction.  I'd be curious to hear other perspectives.

**Reply by mansellan** [reply]
@mwolfe02  Some of those restrictions seem to be from VBx not wanting to deal with the differences between value and reference types, and not wanting to get involved with boxing\unboxing.

It may be that those are solvable with the generic type system of twinBASIC. Is there a reliable list of restrictions for UDTs (ideally with code samples)?
